pkg/apiserver/api_installer.go

```diff
@@ -321,10 +321,12 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 
 			// list or post across namespace.
 			// TODO: more strongly type whether a resource allows these actions on "all namespaces" (bulk delete)
-			namer = scopeNaming{scope, a.group.Linker, gpath.Join(a.prefix, itemPath), true}
-			actions = appendIf(actions, action{"LIST", resource, params, namer}, isLister)
-			actions = appendIf(actions, action{"POST", resource, params, namer}, isCreater)
-			actions = appendIf(actions, action{"WATCHLIST", "watch/" + resource, params, namer}, allowWatchList)
+			if !hasSubresource {
+				namer = scopeNaming{scope, a.group.Linker, gpath.Join(a.prefix, itemPath), true}
+				actions = appendIf(actions, action{"LIST", resource, params, namer}, isLister)
+				actions = appendIf(actions, action{"POST", resource, params, namer}, isCreater)
```

Comment:
----- Original Message -----
> > @@ -321,10 +321,12 @@ func (a *APIInstaller) registerResourceHandlers(path
> > string, storage rest.Storag
> >  
> >  			// list or post across namespace.
> >  			// TODO: more strongly type whether a resource allows these actions on
> >  			"all namespaces" (bulk delete)
> > -			namer = scopeNaming{scope, a.group.Linker, gpath.Join(a.prefix,
> > itemPath), true}
> > -			actions = appendIf(actions, action{"LIST", resource, params, namer},
> > isLister)
> > -			actions = appendIf(actions, action{"POST", resource, params, namer},
> > isCreater)
> > -			actions = appendIf(actions, action{"WATCHLIST", "watch/" + resource,
> > params, namer}, allowWatchList)
> > +			if !hasSubresource {
> > +				namer = scopeNaming{scope, a.group.Linker, gpath.Join(a.prefix,
> > itemPath), true}
> > +				actions = appendIf(actions, action{"LIST", resource, params, namer},
> > isLister)
> > +				actions = appendIf(actions, action{"POST", resource, params, namer},
> > isCreater)
> 
> Do we need cross-namespace POST for anything? If not, we should remove it. If
> so, I think it needs to be distinguished from LIST and WATCH.

We use cross namespace POST for our ResourceAccessReview and SubjectAccessReview endpoints in OpenShift - where you ask the server whether you have permission to do something.  Those are modeled using apiserver/installer today, and they apply to both cluster and non-cluster scoped resources.

> 
> The fully a la carte approach is flexible, but makes it kind of hard to
> understand what combinations of operations we do/should actually support.
> 
> We have only a handful of combinations that make sense (maybe not 100%
> exhaustive, but close, I think, esp. once we eliminate the legacy
> prefix-based verbs: /bindings, /proxy/..., /watch/...):
> * Non-namespaced collections
> * Namespaced collections
> * Subresources for special mutations
> * Subresources for proxying
>   * full HTTP proxying
>   * restricted proxing
> 
> It might be worthwhile to define some "profiles" for these common patterns.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/9991/files#r32760114